backends/graphics/dispmanxsdl/dispmanxsdl-graphics.cpp

```diff
@@ -0,0 +1,589 @@
+		// wait until a free page is freed by vsync CB.
+		if (!page) {
+			pthread_mutex_lock(&_dispvars->vsyncCondMutex);
+			pthread_cond_wait(&_dispvars->vsyncCondition, &_dispvars->vsyncCondMutex);
+			pthread_mutex_unlock(&_dispvars->vsyncCondMutex);
+		}
+	}
+
+	// We mark the choosen page as used
+	pthread_mutex_lock(&page->pageUsedMutex);
+	page->used = true;
+	pthread_mutex_unlock(&page->pageUsedMutex);
+
+	return page;
+}
+
+void DispmanXSdlGraphicsManager::dispmanXFreeResources(void) {
+	// What if we run into the vsync cb code after freeing the resources?
+	pthread_mutex_lock(&_dispvars->pendingMutex);
+	if (_dispvars->pageflipPending > 0)
+	{
+	     pthread_cond_wait(&_dispvars->vsyncCondition, &_dispvars->pendingMutex);
+	}
+	pthread_mutex_unlock(&_dispvars->pendingMutex);
+
+	for (int i = 0; i < _dispvars->numpages; i++) {
+		vc_dispmanx_resource_delete(_dispvars->pages[i].resource);
+		_dispvars->pages[i].resource = 0;
+		_dispvars->pages[i].used = false;
+	}
+
+	_dispvars->update = vc_dispmanx_update_start(0);
+	vc_dispmanx_element_remove(_dispvars->update, _dispvars->element);
+	vc_dispmanx_update_submit_sync(_dispvars->update);
+	// We use this on the setup function to know if we have to free resources and element.
+	_dispvars->element = 0;
+}
+
+void DispmanXSdlGraphicsManager::dispmanXVideoQuit() {
+	// This also waits for pending flips to complete, that's needed before
+	// we destroy the mutexes and condition.
+	dispmanXFreeResources();
+
+	// Destroy the mutexes and conditions
+	for (int i = 0; i < _dispvars->numpages; i++) {
+		pthread_mutex_destroy(&_dispvars->pages[i].pageUsedMutex);
+	}
+	pthread_mutex_destroy(&_dispvars->pendingMutex);
+	pthread_mutex_destroy(&_dispvars->vsyncCondMutex);
+	pthread_cond_destroy(&_dispvars->vsyncCondition);
+
+	free(_dispvars->pages);
+
+	// Close display and deinit
+	vc_dispmanx_display_close(_dispvars->display);
+	bcm_host_deinit();
+}
+
+bool DispmanXSdlGraphicsManager::loadGFXMode() {
+	_forceFull = true;
+
+	// In dispmanX, we manage aspect ratio correction, so for scummvm it's always disabled.
+	_videoMode.aspectRatioCorrection = false;
+
+	_videoMode.overlayWidth = _videoMode.screenWidth;
+	_videoMode.overlayHeight = _videoMode.screenHeight;
+	_videoMode.hardwareWidth = _videoMode.screenWidth;
+	_videoMode.hardwareHeight = _videoMode.screenHeight;
+
+	//
+	// Create the surface that contains the 8 bit game data
+	//
+
+	_screen = SDL_CreateRGBSurface(SDL_SWSURFACE, _videoMode.screenWidth, _videoMode.screenHeight,
+		_screenFormat.bytesPerPixel << 3,
+		((1 << _screenFormat.rBits()) - 1) << _screenFormat.rShift ,
+		((1 << _screenFormat.gBits()) - 1) << _screenFormat.gShift ,
+		((1 << _screenFormat.bBits()) - 1) << _screenFormat.bShift ,
+		((1 << _screenFormat.aBits()) - 1) << _screenFormat.aShift );
+		if (_screen == NULL)
+			error("allocating _screen failed");
+	// Avoid having SDL_SRCALPHA set even if we supplied an alpha-channel in the format.
```

Comment:
Something is wrong with the indention here; the `if` statement seems to be one level too deep?